ludwig/hyperopt/execution.py

```diff
@@ -806,44 +787,73 @@ def run_experiment_trial(config, local_hyperopt_dict, checkpoint_dir=None):
 
         run_experiment_trial_params = tune.with_parameters(run_experiment_trial, local_hyperopt_dict=hyperopt_dict)
 
+        if _is_ray_backend(backend):
+            # If Ray backend, only request custom resource at trial level (inner Tuner will request resources)
+            resources = [self.trial_function_resources]
+        else:
+            # If not Ray backend, request all of the resources required at the trial level
+            use_gpu = bool(self._gpu_resources_per_trial_non_none)
+            num_cpus, num_gpus = _get_num_cpus_gpus(use_gpu)
+            resources = [self.trial_function_resources, {"CPU": num_cpus, "GPU": num_gpus}]
```

Comment:
If we have a single bundle, then we can only schedule the trials on nodes with all three resources. Here, we ensure that the trial driver is scheduled on the head node, and that the worker can be scheduled wherever it needs to be.

Reply:
@geoffreyangus Seems worth adding this as a comment in the code.

Reply:
@geoffreyangus for local execution mode, the trial driver is the worker, so I think you would just want the one bundle.

Reply:
Right that makes sense– I've added a couple of comments to this code block that explains the contents of the `resources` list. Copied here for convenience:

```
if _is_ray_backend(backend):
    # NOTE(geoffrey): as of PR #2709, when we are using a Ray backend, there are two main processes
    # spawned during a Ray Tune trial.
    #
    # The first process is the `trial_function`, which is the function that is passed to `tune.run`.
    # The `tuner` object allocates resources for this function using `self.trial_function_resources` below.
    # Note that `trial_function` is a very lightweight process whose main purpose is to spawn the `trainer`
    # process and communicate with the `tuner` object.
    #
    # The second process is the actual `trainer`, which is spawned by `trial_function`. The `tuner` object
    # does not know about this process, so we cannot allocate resources for it ahead of time. Instead, the
    # `trainer` will reserve its own resources downstream based on its config.
    #
    # In summary, we should only request resources for `trial_function` at this level.
    # `trainer` will request its own resources.
    resources = [self.trial_function_resources]
else:
    # If we are NOT using the Ray backend, the `trial_function` and the `trainer` are executed in the same
    # process. Since the `trial_function` doesn't really require its own resources, we can just request
    # resources for the `trainer` here.
    use_gpu = bool(self._gpu_resources_per_trial_non_none)
    num_cpus, num_gpus = _get_num_cpus_gpus(use_gpu)
    resources = [{"CPU": num_cpus, "GPU": num_gpus}]
```

Reply:
For local execution mode, I initially requested resources for `self.trial_function_resources` in addition to the resources required for the trainer to ensure consistent behavior between the two execution modes. 

In the latest change, I've removed `self.trial_function_resources` in the above code because you are right, it isn't explicitly necessary to request those additional resources in local execution mode. That said, we can't put `self.trial_function_resources` in the same bundle as the other trainer resources. If we did that, we would restrict task placement to pods with all three resource types (trial_function_resources, CPUs, GPUs), which is not necessarily desirable.

Reply:
Updated the above comment to use the terms defined here: https://github.com/ludwig-ai/ludwig/pull/2709#discussion_r1062880369